### PR TITLE
docs/esp8266: Add APA102 tutorial.

### PR DESCRIPTION
**Pull request by mcauser**

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for the submission, and sorry for the slow response.

It looks good, and I'm happy to have this in the tutorial, except there's one issue: the example code doesn't work (at least for me) because the APA102 class is created by default with `bpp=4` (including intensity) yet the colours written to it only have 3 elements in the tuple.

So there probably should be some clarification in the intro of this tutorial about whether the APA102 needs intensity or not, how the user should select it themselves (ie just by specifying the `bpp` keyword), and then update to the demo so it works.

Thanks!

File: docs/esp8266/tutorial/apa102.rst [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            return (0, offset * 3, 255 - offset * 3)
+        offset -= 170
+        return (offset * 3, 255 - offset * 3, 0)
+
+    # Demo 1: RGB
+    red = 0xff0000
+    green = red >> 8
+    blue = red >> 16
+    for i in range(strip.n):
+        color = red >> (i % 3) * 8
```
Comment (dpgeorge) [inline docs/esp8266/tutorial/apa102.rst]: Probably should be `colour` for consistency

Comment (mcauser) [reply docs/esp8266/tutorial/apa102.rst]: _I've been writing too much css_

**Comment by mcauser** [comment]
It's been a while since I've worked with these strips. I'll recheck all of the code examples and amend

**Comment by dpgeorge** [comment]
According to the quickref, it looks like they need an intensity entry as the 4th element of the tuple.

**Comment by mcauser** [comment]
Upgraded and tested examples with ESP8266 + 1M APA102C strip.
Added instructions for changing tuple colour order for compatibility with APA102 + APA102C.
My APA102C uses RBG instead of RGB as with the APA102.

IMG_1098b

**Comment by mcauser** [comment]
The LED driver expects 4 bpp. R,G,B,Brightness (or R,B,G,Brightness).

The default value for brightness appears to be 0.
```
import machine, apa102
strip = apa102.APA102(machine.Pin(5), machine.Pin(4), 5)
strip.bpp = 3
strip[0] = (255, 255, 255)
strip.write()
```
LED strip does not light up - expected when brightness=0.

```
strip.bpp = 4
strip[0] = (255, 255, 255, 7)
strip.write()
```
LED strip shows a white pixel.

```
strip.bpp = 3
strip[0] = (255, 0, 0)
strip.write()
```
LED strip shows a red pixel, at the previous brightness=7.
The previous brightness 4th element in the tuple persists, but you can't see/change it.

```
strip.bpp = 4
print(strip[0])
(255, 0, 0, 7)
```
Set it back to 4 and you can see/change it again.

**Comment by dpgeorge** [comment]
> The default value for brightness appears to be 0.

That would be because the `bytearray()` that's used as the buffer is initialised to all 0's.  This array is (R,G,B,I,R,G,B,I,...) corresponding to the entire strip, and the I's start out as 0 intensity (along with all the R, G, B).  If you change an I then it'll stay that way until it's changed again.

But the point is that all APA102 require 4 bytes per pixel to be sent out, and it just depends how those 4 bytes are set.

**Comment by dpgeorge** [comment]
Ok, merged in 6323cbda4fd82f2396a19e18a224a2579a392303 with a minor change to the paragraph about the RGB byte order.

**Pull request closed without merging**